message.go

```diff
@@ -845,6 +909,43 @@ func (m *Message) unmarshal(v interface{}) error {
 	return nil
 }
 
+func (m *Message) unmarshalToMap(rv reflect.Value) error {
+	for _, k := range m.keys {
+
+		value, ok := m.data[k]
+		if !ok {
+			continue
+		}
+
+		// Map Value Kind
+		mapElemType := rv.Type().Elem()
+		switch mapElemType.Kind() {
```

Comment:
`rv` is the map here isn't it? So isn't this switch statement always going to follow the `reflect.Map` case?

Shouldn't you be checking the type of `k` instead? And what about a check to make sure `value` is assignable to this map?

Reply:
I added a check for the key type.

I also re-wrote this section a bit as it still had a few edge cases it was failing (added new test cases as well).

Basically what is happening is that a value from `MapIndex` cannot be set directly since they are not addressable (also they're Invalid if the key doesn't exist, despite what the docs say about returning a zero value). So we have to use reflect to create a new instance of the map value, set that, and then use `SetMapIndex` to put it into the map.